R/shared-methods.R

```diff
@@ -509,6 +509,158 @@ DataClass <- R6::R6Class(
+    #' run through each class.
+    #' @param download logical. To download the data (TRUE) or use a snapshot
+    #' (FALSE). Defaults to FALSE.
+    #' @importFrom purrr walk map
+    #' @importFrom dplyr slice_tail
+    #' @importFrom testthat test_that expect_s3_class expect_true expect_error
+    test = function(download = FALSE) {
+      data_name <- paste0(class(self)[1], " at level ", self$level)
+      raw_path <- paste0(
+        "custom_data/", class(self)[1],
+        "_level_", self$level, ".rds"
+      )
+      if (!file.exists(raw_path)) {
+        download <- TRUE
+      }
+      if (download) {
+        test_that(paste0(data_name, " downloads sucessfully"), {
+          self$download()
+          walk(self$data$raw, function(data) {
+            expect_s3_class(data, "data.frame")
+            expect_true(nrow(data) > 0)
+            expect_true(ncol(data) >= 2)
+          })
+        })
+        self$data$raw <- map(self$data$raw,
+          slice_tail,
+          n = 250
+        )
+        self$data$raw <- map(
+          self$data$raw,
+          ~ .[, 1:min(100, ncol(.))]
+        )
+        saveRDS(self$data$raw, raw_path)
+      } else {
+        self$data$raw <- readRDS(raw_path)
+      }
+      test_that(paste0(data_name, " can be cleaned as expected"), {
+        self$clean()
+        expect_s3_class(self$data$clean, "data.frame")
+        expect_true(nrow(self$data$clean) > 0)
+        expect_true(ncol(self$data$clean) >= 2)
+        self$expect_clean_cols(self$data$clean)
+      })
+
+      test_that(
+        paste0(data_name, " can highlight available regions as expected"),
+        {
+          expect_error(self$available_regions(), NA)
+          expect_true(class(self$available_regions()) %in% "character")
+        }
+      )
+
+      test_that(paste0(data_name, " can be processed as expected"), {
+        self$process()
+        expect_s3_class(self$data$processed, "data.frame")
+        expect_true(nrow(self$data$processed) > 0)
+        expect_true(ncol(self$data$processed) >= 2)
+        self$expect_processed_cols(self$data$processed, level = self$level)
+        if (!class(self)[1] %in% c("ECDC", "WHO")) {
+          local_region <- self$clone()
+          local_region$localise <- FALSE
+          local_region$process()
+          self$expect_processed_cols(
+            local_region$data$processed,
+            level = self$level,
+            localised = FALSE
+          )
+        }
+      })
+
+      test_that(paste0(data_name, " can be returned as expected"), {
+        returned <- self$return()
+        if (any(class(returned) %in% "data.frame")) {
+          expect_s3_class(returned, "data.frame")
+          expect_true(nrow(returned) > 0)
+          expect_true(ncol(returned) >= 2)
+        }
+      })
+
+      self$expect_columns_contain_data(data_name)
+    },
+
+    #' @description Expect data has cleaned columns. Inherited by child
```

Comment:
part of me is thinking we want these as separate functions rather than methods. DataClass is getting very full! What do you think?

Reply:
I think a `testing_functions.R` script would be useful and make it more verbose from within testing